### PR TITLE
Added object iterator to allow using ObjectType with queries

### LINKED ISSUE #224: Map for Object Microstates
Do you have any plans to add a `map` transition to Object microstates? I have a collection of values that i want to store in an object keyed by ids so I get O(1) value lookup. However, microstates doesn't seem to currently support functor operations on Objects. I can work around it with a forEach loop, and it's not like the difference of 1 line of code vs 5 is a huge deal. However, microstate code is highly functional in style already, so it would be nice if this was supported too.

<details>
<summary>
Example Code (https://codesandbox.io/s/vv709w87v5)
</summary>

```js
class Todo {
  id = create(String);
  completed = create(Boolean, false);
}

class MyModel {
  todosById = create({ Todo });

  completeAll() {
    let newValue = this;

    // I wish there was a functional way to do this
    // e.g. this.todosById.map(todo => todo.completed.set(true));

    Object.keys(this.todosById.state).forEach(id => {
      newValue = newValue.todosById[id].completed.set(true);
    });

    return newValue;
  }
}
```
</details>

P.S. I tried making a PR to add it myself, but due to my inexperience with functional lenses I couldn't quite make it work. I tried something like the following, but then I end up with each item in the dictionary belonging to it's own world due to the way lenses return the context object. 

```js
// src/types/object.js
  map(fn) {
    return map((value, key) => fn(this[key], key), this.state);
  }
```

I wanted to map over `this` rather than `this.state`, but when I did that I ended up mapping over extra keys like 'state' that aren't actually part of my user land object.

### PR DESCRIPTION
**Pull request by taras**
This PR introduces `Symbol.iterator` for ObjectType. This makes it possible to use object microstates with `for (let obj of microstates)` syntax, which also provides support for Queries to map, filter and reduce objects.

The iterator yields entry objects which allow the key/value to be restructured as objects `for (let { key, value } of microstates)` or arrays `for (let [value key] of microstates)`.

Closes #224

### CODE REVIEW

**Comment by cowboyd** [comment]
@rkeeler suggested in #276 (and my inclination is to agree) that we should be iterating over `key/value` pairs, not just values. In other words, an object is a stream of entries, not just values.

**Comment by taras** [comment]
Yes, I was thinking the same thing. I just wasn't sure what was the best way to do this. Should it be `{key, value}`?

**Comment by taras** [comment]
@cowboyd I made the changes as per our discussion in Slack

**Comment by rkeeler** [comment]
I like that you can destructure the pairs in the for/of loop as either an array or an object. But I think the more common case over using for/of is using an anonymous function with the microstate queries. And in that case, my preference for the API would be making the key and value be separate arguments to the supplied function so that you don't have to deal with destructuring and can leave the key/index out most of the time. Granted it's a small sample size, but I've actually yet to need the key or index in any of the queries i've written. The majority look like this.

```js
get player() {
  return userName => find(this.players, p => p.userName.state === userName);
}
```

**Comment by rkeeler** [comment]
Actually, looking into it a bit more. I think my preferred API for exposing the key/index to the query functions can still be achieved and isn't hindered by the implementation in this PR. It would be an additional change to the implementation of the query functions.

I had to add a few new queries for what i was doing (such as find and forEach) and I'll be making a PR with those soon. I could throw in changes to expose the keys and indexes to the queries in that PR.

File: src/query.js [hunk 1]
```diff
@@ -14,6 +14,17 @@ export function reduce(iterable, fn, initial) {
...
 }
 
+export class Entry {
+  constructor(key, value) {
+    this.key = key;
+    this.value = value;
+  }
+
+  [Symbol.iterator]() {
+    return [this.value, this.key][Symbol.iterator]();
```
Comment (cowboyd) [inline src/query.js]: Sometimes you come across a line of code that is just pleasing to look at. This is one of those lines.

File: src/types/object.js [hunk 2]
```diff
@@ -2,6 +2,7 @@ import { append, filter, map } from 'funcadelic';
 import parameterized from '../parameterized';
 import { mount, valueOf } from '../meta';
 import { create } from '../microstates';
+import { Entry } from '../query';
```
Comment (cowboyd) [inline src/types/object.js]: `Entry` is a concept of `Object`, and not `Query`, and so should probably reside in the object class.

**Comment by cowboyd** [comment]
@rkeeler the Microstates query mechanism is actually drawn from a proof of concept that I through together here https://github.com/cowboyd/generation.js/blob/master/src/query.js

I'd like to eventually use this library instead since there's nothing really microstates about the query mechanism. It's really just comprehensions over iterables. 

`find` would make a welcome addition, and I'd also like to get `union`, `intersect`, `disjunct`, and friends in there as well.

File: src/query.js (hunk 1 above)
Comment (taras) [reply src/query.js]: 🤗

**Comment by taras** [comment]
@cowboyd I've made all of the changes.

**Comment by cowboyd** [comment]
Can we make this a single merge commit?

**Pull request merged**